File: dotnet/VedaException.cs
namespace VedaDB;

/// <summary>
/// Base exception for all VedaDB errors.
/// </summary>
public class VedaException : Exception
{
    public VedaException(string message) : base(message) { }
    public VedaException(string message, Exception inner) : base(message, inner) { }
}

/// <summary>
/// Thrown when a connection to VedaDB fails or is lost.
/// </summary>
public class VedaConnectionException : VedaException
{
    public VedaConnectionException(string message) : base(message) { }
    public VedaConnectionException(string message, Exception inner) : base(message, inner) { }
}

/// <summary>
/// Thrown when VedaDB returns a query error.
/// </summary>
public class VedaQueryException : VedaException
{
    public VedaQueryException(string message) : base(message) { }
}

/// <summary>
/// Thrown when a circuit breaker is open.
/// </summary>
public class VedaCircuitBreakerOpenException : VedaException
{
    public VedaCircuitBreakerOpenException(string message) : base(message) { }
}

/// <summary>
/// Thrown when all retries are exhausted.
/// </summary>
public class VedaRetryExhaustedException : VedaException
{
    public VedaRetryExhaustedException(string message) : base(message) { }
    public VedaRetryExhaustedException(string message, Exception inner) : base(message, inner) { }
}

/// <summary>
/// Thrown when a failover operation fails.
/// </summary>
public class VedaFailoverException : VedaException
{
    public VedaFailoverException(string message) : base(message) { }
    public VedaFailoverException(string message, Exception inner) : base(message, inner) { }
}

/// <summary>
/// Thrown when the connection pool is exhausted.
/// </summary>
public class VedaPoolExhaustedException : VedaException
{
    public VedaPoolExhaustedException(string message) : base(message) { }
}

/// <summary>
/// Thrown when a bulk insert operation fails.
/// </summary>
public class VedaBulkInsertException : VedaException
{
    public VedaBulkInsertException(string message) : base(message) { }
    public List<int>? FailedBatchIndices { get; init; }
}

/// <summary>
/// Thrown when a URI parse operation fails.
/// </summary>
public class VedaUriParseException : VedaException
{
    public VedaUriParseException(string message) : base(message) { }
}

/// <summary>
/// Thrown when an interceptor rejects an operation.
/// </summary>
public class VedaInterceptorException : VedaException
{
    public VedaInterceptorException(string message) : base(message) { }
}

File: php/src/VedaException.php
<?php

declare(strict_types=1);

namespace VedaDB;

/**
 * Base exception for all VedaDB driver errors.
 */
class VedaException extends \RuntimeException
{
    private ?string $sqlState;
    private ?int $errorCode;

    public function __construct(
        string $message = '',
        ?int $errorCode = null,
        ?string $sqlState = null,
        ?\Throwable $previous = null
    ) {
        parent::__construct($message, 0, $previous);
        $this->errorCode = $errorCode;
        $this->sqlState = $sqlState;
    }

    public function getErrorCode(): ?int
    {
        return $this->errorCode;
    }

    public function getSqlState(): ?string
    {
        return $this->sqlState;
    }
}

/**
 * Thrown on TCP connection failures.
 */
class ConnectionException extends VedaException {}

/**
 * Thrown when the server returns a query error.
 */
class QueryException extends VedaException {}

/**
 * Thrown on socket timeout.
 */
class TimeoutException extends VedaException {}

/**
 * Thrown on authentication failures.
 */
class AuthException extends VedaException {}

/**
 * Thrown when the circuit breaker is open.
 */
class CircuitOpenException extends VedaException {}

/**
 * Thrown on failover exhaustion.
 */
class FailoverException extends VedaException {}

/**
 * Thrown on pool exhaustion.
 */
class PoolExhaustedException extends VedaException {}

/**
 * Thrown on TLS/SSL errors.
 */
class TLSSException extends VedaException {}

/**
 * Thrown on protocol errors.
 */
class ProtocolException extends VedaException {}

/**
 * Thrown on validation errors.
 */
class ValidationException extends VedaException {}
